### ISSUE TITLE
Android deeplinks data not available when using with FCM

### ISSUE THREAD

**Issue by Imgkl** [customerio/customerio-flutter#31]
I am able to receive the notification, but the custom data is lost.

**Comment by mrehan27** [comment]
Hi @Imgkl. Sorry to hear that. We unfortunately don't yet expose any option in Flutter SDK to read custom data from notifications received yet. However, the data should still not be lost and be available in notification extras. Can you please share code snippet you are using to read it so we might be able to help?

**Comment by Imgkl** [comment]
Sure, @mrehan27 .

This is the custom data that's I'm trying to send.

````
{   
    "fromCustomerIO": "true",
    "redirect": "search",
    "triggerString": "AAPL"
}
````

This code is what responsible to listen and read notifications.

````
   FirebaseMessaging.instance
        .getInitialMessage()
        .then((message) => _processNotification(message));

    FirebaseMessaging.onMessageOpenedApp
        .listen((RemoteMessage message) => _processNotification(message));

    FirebaseMessaging.onBackgroundMessage(
        (message) => _processNotification(message));
````

I have also added this code snippet which I saw in the documentation.

````
<service
    android:name="io.customer.messagingpush.CustomerIOFirebaseMessagingService"
    android:exported="false">
    <intent-filter>
        <action android:name="com.google.firebase.MESSAGING_EVENT" />
    </intent-filter>
</service>
````

**Comment by mrehan27** [comment]
Thanks for all the details @Imgkl. I belive the classes you are using are from [Firebase Messaging Plugin for Flutter](https://pub.dev/packages/firebase_messaging). Can you please also confirm if `FirebaseMessaging` events are not being called completely or do you receive callback but with empty `data` only? Also, can you please confirm the behavior for Foreground, Background and Killed/Terminated state as well?

**Comment by Imgkl** [comment]
@mrehan27 That's the right package. 

FirebaseMessaging messaging events are not being called completely. I'm able to receive the notifications on  Foreground, Background and Killed/Terminated state but when I tap on it, The app opens from the start and callback ````messsage```` gives null. 

On iOS, I'm able to get receive the data in the callback. FYI.

**Comment by mrehan27** [comment]
Thanks for the confirmation @Imgkl. It looks like the issue is with multiple notification services. Unfortunately, our Flutter SDK doesn't yet allow passing data between multiple notification services for Android apps. As the [docs suggest](https://www.customer.io/docs/sdk/flutter/push/#multiple-push-services), adding `CustomerIOFirebaseMessagingService` will take away control from other notification libraries. We are aware of this challenge and is something in our roadmap. We haven't yet started working on it, but I can add your vote to it so we can notify when it is available.

Meanwhile, to make rich push notifications and custom data work together, you can try following steps:

- Remove `CustomerIOFirebaseMessagingService` from your `AndroidManifest`
- Add listeners to `FirebaseMessaging.onBackgroundMessage` and `FirebaseMessaging.onMessage`
  - Handle [metrics tracking](https://www.customer.io/docs/sdk/flutter/push/#metrics-tracking) to make sure your delivered/opened metrics are not affected
  - Send local notification to device
  - Handle your custom data

I do apologize for any inconvenience caused due to this. Please let us know if that doesn't work for you or if you have more questions.

**Comment by Imgkl** [comment]
@mrehan27 I'll try this and update you the results.

**Comment by Imgkl** [comment]
@mrehan27 after removing the ````CustomerIOFirebaseMessagingService```` from the manifest file, I'm not even able to receive notifications sent from campaigns dashboard.

**Comment by mrehan27** [comment]
@Imgkl If you just removed `CustomerIOFirebaseMessagingService` from the manifest file, you'll not receive any notifications from Customer.io SDK. However, you should receive callbacks from `FirebaseMessaging` class with expected payload in `data` object and should then be able to build notifications by yourself.

For clarity, once you remove `CustomerIOFirebaseMessagingService` from the manifest file, Customer.io SDK will not be able to receive and/or handle any notifications received. But the dashboard behavior and notification payloads are unaffected, so you can receive and handle the notifications by yourself. This obviously is not ideal, and we do have in our plan to provide more callbacks so Customer.io SDK can build notifications for you. But since that is being prioritized and we haven't started working on it yet, the suggested workaround should unblock you till we roll out these improvements in our SDK. Thank you for the patience.

**Comment by mingjunsiek** [comment]
@mrehan27 I am also facing this issue and I was able to resolve it by adding the `CustomerIOFirebaseMessagingService` service in the `AndroidManifest` file. However, I am facing a weird side effect on android. Even though the app is in the background, when I tapped on the push notification, it always instantiate a new instance of the app, even though it should have continued the activity on the background. Could you advise?

**Comment by Imgkl** [comment]
This happens for me as well.

**Comment by Imgkl** [comment]
@mrehan27 Yes, Thanks for the temp workaround, but it's not ideal in my use case. Guess we are going to wait till a proper fix is rolled out. 

Feel free to close this issue.

**Comment by mrehan27** [comment]
@Imgkl That's understandable. Thank you for the patience.

Keeping the issue open till we roll out the fix so anyone facing similar issue can track the progress here.

**Comment by mingjunsiek** [comment]
@mrehan27 can u further explain on how to do this?

> Add listeners to FirebaseMessaging.onBackgroundMessage and FirebaseMessaging.onMessage
Handle [metrics tracking](https://www.customer.io/docs/sdk/flutter/push/#metrics-tracking) to make sure your delivered/opened metrics are not affected
Send local notification to device
Handle your custom data

FirebaseMessaging.onBackgroundMessage spawns a new isolate so I could not use the tracking controller and CIO controller created in the main isolate to send metrics over to CIO and Amplitude...

**Comment by mingjunsiek** [comment]
@mrehan27, it seems that there is also some [issue](https://github.com/firebase/flutterfire/issues/9718) with FlutterFire FirebaseMessaging as well. This means we won't be able to send the metric for `Delivered` and `Opened` back to CIO on iOS. Do u have any suggestions for this??

**Comment by mingjunsiek** [comment]
@mrehan27 there seems to also be an issue for iOS side regarding the 'delivered' metric.
In this screenshot, u can see 2 push notif that is received on the device. I've opened up the last notif. As u can see on the right, the last notif's status has been set to 'Opened' but the first to is still stuck at 'Sent' even though they were already delivered. I was using the Rich Push notification on iOS for sending this notif. Any idea why?

![image](https://user-images.githubusercontent.com/13094556/226804098-47882ba4-522b-4c4b-a2a8-a2b2c658e7f1.png)

**Comment by FDuhen** [comment]
> @Imgkl That's understandable. Thank you for the patience.
> 
> Keeping the issue open till we roll out the fix so anyone facing similar issue can track the progress here.

Having the same kind of issue when trying to use Local Notifications with CustomerIO's SDK added to the project.

**Comment by mrehan27** [comment]
We have released more changes in [`1.1.0`](https://github.com/customerio/customerio-flutter/releases) that makes it easier to receive/process notifications with [Firebase Messaging Plugin for Flutter](https://pub.dev/packages/firebase_messaging). You can read more [details in our docs](https://www.customer.io/docs/sdk/flutter/push/#multiple-push-services).

Closing this as the update solves the problem of integrating and reading notification data using [Firebase Messaging Plugin for Flutter](https://pub.dev/packages/firebase_messaging). If you have any questions or any feedback to share, please don't hesitate to reach us.

**Comment by eriquezinedine** [comment]
### Enriched notifications are working correctly on both Android and iOS.

However, on Android, tapping the notification does not navigate as expected. Instead of using the deep_link field from the Customer.io form, I tried using the custom data (data object) directly, but I’m getting the error shown in the attached image.

While reviewing the Customer.io documentation in the Android (native) section —not the Flutter one— I found a possible solution that involves adding Kotlin code to properly handle the click action and parse the notification data.

From what I understand, this seems to be the correct solution. Or do you have any other suggestions to solve this issue?


<img width="1366" height="1316" alt="Image" src="https://github.com/user-attachments/assets/7e92c6f7-7c8f-4be6-986d-84297a7e8db9" />

Error.

<img width="934" height="327" alt="Image" src="https://github.com/user-attachments/assets/e961d668-9dcf-45f4-a87b-c52b65ad6a29" />
